sql/database_test.go

```diff
@@ -167,47 +169,321 @@ func TestDatabaseSkipMigrations(t *testing.T) {
+	migration1 := &sqlMigration{
+		order:   1,
+		name:    "0001_initial.sql",
+		content: "create table foo(x int)",
+	}
+	migration2 := &faultyMigration{
+		panic: true,
+		sqlMigration: &sqlMigration{
+			order:   2,
+			name:    "0002_test.sql",
+			content: "create table bar(y int)",
+		},
+	}
+
+	dbFile := filepath.Join(dir, "test.sql")
+	db, err := Open("file:"+dbFile,
+		WithLogger(logger),
+		WithDatabaseSchema(&Schema{
+			Script: "PRAGMA user_version = 1;\n" +
+				"CREATE TABLE foo(x int);\n",
+			Migrations: MigrationList{migration1},
+		}),
+		WithForceMigrations(true),
+	)
+	require.NoError(t, err)
+	require.NoError(t, db.Close())
+
+	schema := &Schema{
+		Script: "PRAGMA user_version = 2;\n" +
+			"CREATE TABLE bar(y int);\n" +
+			"CREATE TABLE foo(x int);\n",
+		Migrations: MigrationList{migration1, migration2},
+	}
+
+	func() {
+		defer func() {
+			require.NotNil(t, recover())
+		}()
+		Open("file:"+dbFile,
+			WithLogger(logger),
+			WithDatabaseSchema(schema),
+			WithVacuumState(2),
+		)
+	}()
+
+	// Check that temporary database exists after the simulated crash.
+	// Note that we're checking "*_migrate" not "*_migrate*" to avoid matching
+	// any erroneously created successful migration markers.
+	tmpDBFiles, err := filepath.Glob(filepath.Join(dir, "*_migrate"))
+	require.NoError(t, err)
+	require.NotEmpty(t, tmpDBFiles)
+
+	// Retry migration. The incompletely migrated temporary database should be dropped.
+	migration2.panic = false
+	db, err = Open("file:"+dbFile,
+		WithLogger(logger),
+		WithDatabaseSchema(schema),
+		WithVacuumState(2),
+	)
+	require.NoError(t, err)
+	execSQL(t, db, "select * from foo, bar", -1)
+	require.NoError(t, db.Close())
+}
+
+func TestResumeCopyMigration(t *testing.T) {
+	dir := t.TempDir()
+	logger := zaptest.NewLogger(t)
+	migration1 := &sqlMigration{
+		order:   1,
+		name:    "0001_initial.sql",
+		content: "create table foo(x int)",
+	}
+	// This migration will panic when VACUUM INTO is attempted to copy
+	// the migrated database to the source database location.
+	migration2 := &faultyMigration{
+		interceptVacuumInto: true,
+		sqlMigration: &sqlMigration{
+			order:   2,
+			name:    "0002_test.sql",
+			content: "create table bar(y int)",
+		},
+	}
+
+	dbFile := filepath.Join(dir, "test.sql")
+	db, err := Open("file:"+dbFile,
+		WithLogger(logger),
+		WithDatabaseSchema(&Schema{
+			Script: "PRAGMA user_version = 1;\n" +
+				"CREATE TABLE foo(x int);\n",
+			Migrations: MigrationList{migration1},
+		}),
+		WithForceMigrations(true),
+	)
+	require.NoError(t, err)
+	require.NoError(t, db.Close())
+
+	schema := &Schema{
+		Script: "PRAGMA user_version = 2;\n" +
+			"CREATE TABLE bar(y int);\n" +
+			"CREATE TABLE foo(x int);\n",
+		Migrations: MigrationList{migration1, migration2},
+	}
+
+	func() {
+		defer func() {
+			require.NotNil(t, recover())
+		}()
+		Open("file:"+dbFile,
+			WithLogger(logger),
+			WithDatabaseSchema(schema),
+			WithVacuumState(2),
+		)
+	}()
```

Comment:
Why does `Open` panic? In my opinion we should try to avoid this even with faulty migrations and there is an easier way to assert this:

```suggestion
	require.Panics(t, func() {
		Open("file:"+dbFile,
			WithLogger(logger),
			WithDatabaseSchema(schema),
			WithVacuumState(2),
		)
	})
```

Reply:
Thanks, will fix the assertion. The panic is forced on purpose in the test migration code to simulate system or app crash in the middle of a migration, from which go-spacemesh needs to be able to recover upon next startup

Reply:
Switched to `require.Panics`